docs/source/conceptual/flash_attention.md

```diff
@@ -0,0 +1,7 @@
+# Flash Attention
+
+Scaling transformer architecture is heavily bottlenecked by the self-attention mechanism, which has quadratic time and memory complexity. Recent developments in accelerator hardware are mainly focused on enhancing compute capacities and not memory and transferring data between hardware. This results in attention operation having a bottleneck in memory, also called _memory-bound_. Flash Attention is an attention algorithm used to overcome this problem and scale transformer-based models more efficiently, enabling faster training and inference. 
+In standard attention implementation, the cost of loading and writing keys, queries, and values from High Bandwidth Memory (HBM) is high. It loads key, query, value from HBM to GPU, performs a single step of the attention mechanism and writes it back to HBM, and repeats this for every singular step of the attention. Instead, Flash Attention loads keys, queries, and values once, fuses the operations of the attention mechanism and writes them back. 
```

Comment:
```suggestion
In standard attention implementation, the cost of loading and writing keys, queries, and values from High Bandwidth Memory (HBM) is high. It loads key, query, value from HBM to GPU on-chip SRAM., performs a single step of the attention mechanism and writes it back to HBM, and repeats this for every singular step of the attention. Instead, Flash Attention loads keys, queries, and values once, fuses the operations of the attention mechanism and writes them back. 
```

Both rams are on the GPU, but there's the global address RAM, and the chip local (L1 equivalent if you will) RAM.

Reply:
Maybe even link the paper directly from more complete info ? https://arxiv.org/pdf/2205.14135.pdf

Reply:
I made this previously to clarify, WDYT? ![image](https://github.com/huggingface/text-generation-inference/assets/53175384/824ba962-43b0-4834-b94a-cbfdfff639d8)

Reply:
Just link the paper, it's complex enough I think :)

Reply:
The main thing is that it doesn't "solve" the quadratic issue, it just makes the operation more efficient on the hardware.